### PR TITLE
Add TT-Dit SD3.5 CLIP & T5

### PR DESCRIPTION
**Pull request by DorsaRoh**
### Ticket
N/A

### What's changed
Adds TT-DiT SD3.5 CLIP and T5 encoders

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes

### CODE REVIEW

File: models/experimental/tt_dit/layers/feedforward.py [hunk 1]
```diff
@@ -37,8 +37,18 @@ def __init__(
...
+        has_ff_keys = any(k.startswith("ff1.") or k.startswith("ff2.") for k in state_dict.keys())
+
+        if has_fc_keys:
+            # CLIP format: fc1, fc2
+            self.ff1.load_state_dict(substate(state_dict, "fc1"))
+            self.ff2.load_state_dict(substate(state_dict, "fc2"))
+        else:
+            # standard format: ff1, ff2
+            self.ff1.load_state_dict(substate(state_dict, "ff1"))
+            self.ff2.load_state_dict(substate(state_dict, "ff2"))
```
Comment (DorsaRoh) [inline models/experimental/tt_dit/layers/feedforward.py]: Are there better dynamic weight key handling alternatives?

**Review by mbahnasTT — APPROVED** [review]

File: models/experimental/tt_dit/encoders/t5/model_t5.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+    def load_state_dict(self, state_dict):
+        self.wi0.load_state_dict({"weight": state_dict["wi_0.weight"]})
+        self.wi1.load_state_dict({"weight": state_dict["wi_1.weight"]})
+        self.wo.load_state_dict({"weight": state_dict["wo.weight"]})
+
+    def __call__(self, x: ttnn.Tensor) -> ttnn.Tensor:
+        gelu = new_gelu_activation(self.wi0(x))
+        linear = self.wi1(x)
+        x = gelu * linear
+        hidden_states = self.wo(x)
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/encoders/t5/model_t5.py:228]: Since wo is col-parallel, would you need an all-gather on x before `self.wo(x)`?

File: models/experimental/tt_dit/layers/feedforward.py (hunk 1 above)
Comment (cglagovichTT) [reply models/experimental/tt_dit/layers/feedforward.py]: I'd prefer to remap the key names in `CLIPEncoderLayer.load_state_dict` and leave feedforward `load_state_dict` unchanged

File: models/experimental/tt_dit/encoders/t5/model_t5.py [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+        )
+        self.wi1 = ColParallelLinear(
+            in_features=self.config.embed_dim,
+            out_features=self.config.ff_dim,
+            bias=False,
+            mesh_device=self.mesh_device,
+            mesh_axis=self.parallel_config.tensor_parallel.mesh_axis,
+            init=True,
+        )
+        self.wo = ColParallelLinear(
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/encoders/t5/model_t5.py]: I believe this should be RowParallelLinear if you're passing `shard_dim=-2`

File: models/experimental/tt_dit/layers/linear.py [hunk 4]
```diff
@@ -73,20 +73,35 @@ class ColParallelLinear:
...
-        self, in_features, out_features, bias=True, activation=None, mesh_device=None, mesh_axis=0, init=False
+        self,
+        in_features,
+        out_features,
+        bias=True,
+        activation_fn=None,
+        mesh_device=None,
+        mesh_axis=0,
+        init=False,
+        shard_dim=-1,
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/layers/linear.py]: Since this layer is ColParallelLinear, shard dim should not be necessary. For shard_dim=-2, use RowParallelLinear

File: models/experimental/tt_dit/encoders/t5/model_t5.py [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+        if self.parallel_config.tensor_parallel.factor > 1:
+            # hidden_states_scattered = ttnn.experimental.reduce_scatter_minimal_async(
+            #     hidden_states,
+            #     dim=3,
+            #     multi_device_global_semaphore=self.ccl_manager.get_rs_ping_pong_semaphore(),
+            #     num_links=1,
+            #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            #     topology=self.ccl_manager.topology,
+            #     cluster_axis=self.parallel_config.tensor_parallel.mesh_axis,
+            # )
```
Comment (bbradelTT) [inline models/experimental/tt_dit/encoders/t5/model_t5.py]: Should this commented out code be removed?

File: models/experimental/tt_dit/layers/linear.py [hunk 6]
```diff
@@ -128,9 +141,14 @@ def __call__(self, x, core_grid=None, compute_kernel_config=None):
             core_grid=core_grid,
             compute_kernel_config=compute_kernel_config or self.compute_config,
         )
-        if self.activation is not None:
-            assert self.activation == "gelu"
+        if self.activation_fn == "geglu" or self.activation_fn == "gelu":
             output = ttnn.gelu(output, fast_and_approximate_mode=False)
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/layers/linear.py:145]: is geglu really the same as gelu?

File: models/experimental/tt_dit/tests/encoders/clip/test_clip_full_projection.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    clip_path: str,
+    tokenizer_path: str,
+    expected_pcc: float,
+    topology: ttnn.Topology,
+) -> None:
+    parent_mesh_shape = tuple(mesh_device.shape)
+    if any(x[0] < x[1] for x in zip(parent_mesh_shape, submesh_shape)):
+        pytest.skip("submesh shape is larger than parent mesh shape, skipping")
+    encoder_submesh = mesh_device.create_submesh(ttnn.MeshShape(*submesh_shape))
+    print(f"Running on submesh {encoder_submesh.shape} of parent mesh {mesh_device.shape}")
```
Comment (bbradelTT) [inline models/experimental/tt_dit/tests/encoders/clip/test_clip_full_projection.py:58]: Can this be changed to logger?

File: models/experimental/tt_dit/encoders/t5/model_t5.py [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+
+    def __call__(self, x: ttnn.Tensor) -> ttnn.Tensor:
+        # breakpoint()
+        gelu = new_gelu_activation(self.wi0(x))
+        linear = self.wi1(x)
+        x = gelu * linear
+        hidden_states = self.wo(x)
+        hidden_states = ttnn.unsqueeze(hidden_states, 0)
+
+        if self.parallel_config.tensor_parallel.factor > 1:
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/encoders/t5/model_t5.py:231]: col parallel -> row parallel linear means that you need a reduce scatter afterwards

File: models/experimental/tt_dit/pipelines/stable_diffusion_35_large/pipeline_stable_diffusion_35_large.py [hunk 9]
```diff
@@ -241,86 +235,108 @@ def __init__(
         self._image_processor = VaeImageProcessor(vae_scale_factor=self._torch_vae_scale_factor)
 
         # HACK: reshape submesh device 0 to 1D
-        encoder_parallel_manager.mesh_device.reshape(
-            ttnn.MeshShape(*encoder_parallel_manager.tensor_parallel.mesh_shape)
-        )
+        encoder_device.reshape(ttnn.MeshShape(1, 4))  # 1x4 submesh for encoder
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/pipelines/stable_diffusion_35_large/pipeline_stable_diffusion_35_large.py]: Submesh can't be unconditionally reshaped to 1x4. 
Only when the `submesh.shape[1] != 4` and the volume of the submesh is 4 we should reshape to 1x4.

File: models/experimental/tt_dit/pipelines/stable_diffusion_35_large/pipeline_stable_diffusion_35_large.py [hunk 10]
```diff
@@ -493,9 +509,7 @@ def __call__(
 
             with timer.time_section("total_encoding") if timer else nullcontext():
                 # HACK: reshape submesh device 0 from 2D to 1D
-                self.encoder_parallel_manager.mesh_device.reshape(
-                    ttnn.MeshShape(*self.encoder_parallel_manager.tensor_parallel.mesh_shape)
-                )
+                self.encoder_device.reshape(ttnn.MeshShape(1, 4))  # 1x4 submesh for encoder
```
Comment (cglagovichTT) [inline models/experimental/tt_dit/pipelines/stable_diffusion_35_large/pipeline_stable_diffusion_35_large.py]: unconditional reshape

File: models/experimental/tt_dit/encoders/t5/model_t5.py [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+                num_links=self.ccl_manager.num_links,
+                topology=self.ccl_manager.topology,
+                cluster_axis=self.parallel_config.tensor_parallel.mesh_axis,
+                memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            )
+        hidden_states = ttnn.squeeze(hidden_states, 0)
+        return hidden_states
+
+
+def new_gelu_activation(x: ttnn.Tensor) -> ttnn.Tensor:
```
Comment (bbradelTT) [inline models/experimental/tt_dit/encoders/t5/model_t5.py:248]: I wonder if this should be put into a utility function. Probably not as part of this PR.

models/experimental/stable_diffusion_35_large/tt/t5_encoder.py and models/experimental/stable_diffusion_35_large/reference/t5_encoder.py seem to have identical code.

models/experimental/bloom/tt/bloom_gelu_forward.py does something similar, although that use case seems different.
